### ISSUE TITLE
Private/protected properties on siblings in union types

### ISSUE THREAD

**Issue by IceCreamYou** [Microsoft/TypeScript#9974]
**TypeScript Version:**  1.8.9

**Code**

```ts
class Example {
    protected width = 0;
    constructor(sibling: Example | HTMLCanvasElement) {
        console.log(sibling.width);
    }
}
```

**Expected behavior:** No errors.

**Actual behavior:** Error: `Property 'width' does not exist on type 'Example | HTMLCanvasElement'. Line 4, column 31`

The reason this error occurs is that the `width` property is `protected` on `Example` but `public` on `HTMLCanvasElement`. However, the error is confusing for two reasons:

1. The `width` property is available on both component properties in this situation. If the `sibling` parameter had a type of either `Example` or `HTMLCanvasElement`, the code would compile.
2. The error message doesn't explain why the property does not exist on the type. Understanding the error requires the knowledge that properties are only available on union types if they have the same privacy level, even if those properties would be accessible in the component types in the active context.

I am not sure if there are situations in which the desired behavior is a compilation error when accessing a potentially protected property on a union type if that property is accessible on all component types. I can't think of any, and so I think the ideal situation would be that the compiler does not throw an error here. However, I think it would be an adequate solution to improve the error message.

(I apologize if this has been reported/discussed before. I searched but couldn't find anything exactly along these lines.)

**Comment by mhegazy** [comment]
`private` and `protected` properties are tied to their containing class. Whether they exist on the type or not depends on the lexical scope where they are accessed/used. Copying these properties or exposing them on new types (union types) would break these assumptions. 

We would accept PR to enhance the error message.

**Comment by JoshuaKGoldberg** [comment]
> Copying these properties or exposing them on new types (union types) would break these assumptions

@mhegazy Is that a part of the design or just an implementation detail? We can still access the `width` property if we narrow the type:
```typescript
class Example {
    protected width = 0;
    constructor(sibling: Example | HTMLCanvasElement) {
        if (sibling instanceof Example) {
            sibling.width;
        } else {
            sibling.width;
        }
    }
}
```

IMO a better error message would help, but the root issue is that the union types should in fact allow lexical scoping to come into consideration for member accesses. The current behavior is confusing.

**Comment by mhegazy** [comment]
> Is that a part of the design or just an implementation detail? We can still access the width property if we narrow the type:

this is by design. narrowing takes a union type, and removes some constituents. the resulting type(s) is what the variable has withing the block. in these cases, the access to `width` is correct for both types.

> IMO a better error message would help, but the root issue is that the union types should in fact allow lexical scoping to come into consideration for member accesses. The current behavior is confusing.

TS has a structural type system. comparing properties of types happen often,  during assignment compact, but also during narrowing, inference, etc. keeping a "scope" on every property becomes quickly an untraceable problem.
